### PR TITLE
React ui add query stats

### PR DESCRIPTION
**Pull request by CSTDev**
Puts query stats in a useful but unobtrusive place as per #6156

### CODE REVIEW

File: web/ui/react-app/src/QueryStats.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+class QueryStats extends PureComponent<QueryStatsProps>{
+
+    render() {
+        const stats = this.props.stats;
+        const loading = this.props.loading;
+
+        return (
+            <div className="query-stats">
```
Comment (boyskila) [inline web/ui/react-app/src/QueryStats.tsx]: @CSTDev is it needed this component to be created at all if !loading && stats !== null?

File: web/ui/react-app/src/QueryStats.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+interface QueryStatsProps {
+    stats: {
+        loadTime: number;
+        resolution: number;
+        totalTimeSeries: number;
+    } | null;
+    loading: boolean;
+}
+
+class QueryStats extends PureComponent<QueryStatsProps>{
```
Comment (boyskila) [inline web/ui/react-app/src/QueryStats.tsx]: I would suggest you convert it to Functional component since is not maintaining any state or dealing with any of the lifecycles hooks. If you want to preserve it like PureComponent just use memo from React

File: web/ui/react-app/src/Panel.tsx [hunk 3]
```diff
@@ -36,7 +37,11 @@ interface PanelState {
   } | null;
   loading: boolean;
   error: string | null;
-  stats: null; // TODO: Stats.
+  stats: {
```
Comment (boyskila) [inline web/ui/react-app/src/Panel.tsx]: I would suggest you create interface PaneStats or something like that and to use it here and in QueryStats component as well

File: web/ui/react-app/src/Panel.tsx [hunk 4]
```diff
@@ -161,6 +175,11 @@ class Panel extends Component<PanelProps, PanelState> {
           endTime: endTime,
           resolution: resolution,
         },
+        stats: {
+          loadTime: Date.now() - queryStart,
+          resolution: resolution,
```
Comment (boyskila) [inline web/ui/react-app/src/Panel.tsx]: Just resolution is enough, since key: value have same names, same for totalTimeSeries

File: web/ui/react-app/src/QueryStats.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    render() {
+        const stats = this.props.stats;
```
Comment (boyskila) [inline web/ui/react-app/src/QueryStats.tsx]: Just a personal preference. I would destruct the values
const { stats, loading } = this.props
const { loadTime, resolution, totalTimeSeries } = stats

File: web/ui/react-app/src/Panel.tsx [hunk 6]
```diff
@@ -287,7 +307,7 @@ class Panel extends Component<PanelProps, PanelState> {
         <Row>
           <Col>
             <Button className="float-right" color="link" onClick={this.props.removePanel} size="sm">Remove Panel</Button>
-          </Col>
+          </Col>
```
Comment (boyskila) [inline web/ui/react-app/src/Panel.tsx]: unnecessary spaces

File: web/ui/react-app/src/Panel.tsx [hunk 7]
```diff
@@ -153,6 +158,15 @@ class Panel extends Component<PanelProps, PanelState> {
         throw new Error(json.error || 'invalid response JSON');
       }
 
+      let totalTimeSeries = 0;
+      if (json.data !== undefined) {
```
Comment (boyskila) [inline web/ui/react-app/src/Panel.tsx]: I think this will fail if json.data is null

File: web/ui/react-app/src/Panel.tsx [hunk 8]
```diff
@@ -153,6 +158,15 @@ class Panel extends Component<PanelProps, PanelState> {
...
+        if (json.data.resultType === "scalar") {
+          totalTimeSeries = 1;
+        } else if(json.data.result !== null) {
```
Comment (boyskila) [inline web/ui/react-app/src/Panel.tsx]: json.data.result !== null will fail too if json.data.result is undefined

Comment (boyskila) [reply web/ui/react-app/src/Panel.tsx]: Maybe some utility function like isPresent(someValue) will be useful, where isPresent will perform a check for someValue !== null && someValue !== undefined

File: web/ui/react-app/src/QueryStats.tsx (hunk 1 above)
Comment (CSTDev) [reply web/ui/react-app/src/QueryStats.tsx]: Fair point will do the check earlier.

File: web/ui/react-app/src/Panel.tsx (hunk 8 above)
Comment (CSTDev) [reply web/ui/react-app/src/Panel.tsx]: Changed it to if(json.data.result) which should evaluate to false if it's null or undefined.

Comment (boyskila) [reply web/ui/react-app/src/Panel.tsx]: 0 as well :) I think proper check here would (be value !== null && value !== undefined)

File: web/ui/react-app/src/Panel.tsx (hunk 3 above)
Comment (CSTDev) [reply web/ui/react-app/src/Panel.tsx]: Did some renaming, and shared it from QueryStats.

File: web/ui/react-app/src/Panel.tsx (hunk 8 above)
Comment (CSTDev) [reply web/ui/react-app/src/Panel.tsx]: Added a check on length too in case it's an empty object.

File: web/ui/react-app/src/Panel.tsx [hunk 9]
```diff
@@ -157,29 +153,29 @@ class Panel extends Component<PanelProps, PanelState> {
...
-
+      
       let totalTimeSeries = 0;
-      if (json.data !== undefined) {
+      if (json.data) {
         if (json.data.resultType === "scalar") {
           totalTimeSeries = 1;
-        } else if(json.data.result !== null) {
+        } else if(json.data.result && json.data.result.length > 0) {
           totalTimeSeries = json.data.result.length;
```
Comment (boyskila) [inline web/ui/react-app/src/Panel.tsx]: Zero value shouldn't be displayed?

File: web/ui/react-app/src/Panel.tsx [hunk 10]
```diff
@@ -157,29 +153,29 @@ class Panel extends Component<PanelProps, PanelState> {
       if (json.status !== 'success') {
         throw new Error(json.error || 'invalid response JSON');
       }
-
+
```
Comment (boyskila) [inline web/ui/react-app/src/Panel.tsx]: extra spaces :) Not sure wheter is IDE setting or tslint.json should be added to the project to catch such errors. @juliusv

File: web/ui/react-app/src/QueryStatsView.tsx [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import './QueryStatsView.css';
+
+export interface QueryStats{
+    loadTime: number;
+    resolution: number;
+    totalTimeSeries: number;
+}
+
+export default function QueryStatsView(props: QueryStats){
```
Comment (boyskila) [inline web/ui/react-app/src/QueryStatsView.tsx]: I think this should looks like
```
const QueryStatsView = FC<QueryStats> = ({ loadTime, resolution, totalTimeSeries}) => { .... }
```

File: web/ui/react-app/src/QueryStatsView.tsx [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import './QueryStatsView.css';
+
+export interface QueryStats{
```
Comment (boyskila) [inline web/ui/react-app/src/QueryStatsView.tsx]: space before bracket is missing.

File: web/ui/react-app/src/Panel.tsx [hunk 13]
```diff
@@ -157,29 +153,29 @@ class Panel extends Component<PanelProps, PanelState> {
       if (json.status !== 'success') {
         throw new Error(json.error || 'invalid response JSON');
       }
-
+      
       let totalTimeSeries = 0;
-      if (json.data !== undefined) {
+      if (json.data) {
         if (json.data.resultType === "scalar") {
```
Comment (boyskila) [inline web/ui/react-app/src/Panel.tsx]: You can destruct all data values like const { resultType, result } json.data

File: web/ui/react-app/src/Panel.tsx (hunk 9 above)
Comment (CSTDev) [reply web/ui/react-app/src/Panel.tsx]: No they shouldn't and so shouldn't ever get here if there are 0 results, but it's just in case something odd happens, it then displays a 0, which would be correct.

File: web/ui/react-app/src/QueryStatsView.tsx (hunk 11 above)
Comment (CSTDev) [reply web/ui/react-app/src/QueryStatsView.tsx]: Hmm, not see that form before and I can't get anything like it working? No point expanding them out as I created an interface to define them that's then shared too.
Going for:
`const QueryStatsView = (props: QueryStats) => {...}`

Comment (boyskila) [reply web/ui/react-app/src/QueryStatsView.tsx]: Yep, That's fine too

File: web/ui/react-app/src/Panel.tsx (hunk 10 above)
Comment (juliusv) [reply web/ui/react-app/src/Panel.tsx]: Yeah, I'm all for automating this stuff :) https://github.com/prometheus/prometheus/issues/6199

File: web/ui/react-app/src/Panel.tsx [hunk 14]
```diff
@@ -153,14 +154,29 @@ class Panel extends Component<PanelProps, PanelState> {
         throw new Error(json.error || 'invalid response JSON');
```
Comment (juliusv) [inline web/ui/react-app/src/Panel.tsx]: When there's a query error, we should either hide the previously shown query stats or show stats for the errored query (like the old UI does, see http://demo.robustperception.io:9090/graph?g0.range_input=1h&g0.expr=-invalid-&g0.tab=1).

File: web/ui/react-app/src/QueryStatsView.tsx [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    resolution: number;
+    totalTimeSeries: number;
+}
+
+const QueryStatsView = (props: QueryStats) => {
+    const {loadTime, resolution, totalTimeSeries} = props;
+
+    return (
+        <div className="query-stats">
+            <span className="float-right">Load Time: {loadTime}ms &ensp; Resolution: {resolution} &ensp; Total Time Series: {totalTimeSeries}</span>
```
Comment (juliusv) [inline web/ui/react-app/src/QueryStatsView.tsx]: Here's where my specific sense of aesthetics kicks in ;)

Could we change this to:

```suggestion
            <span className="float-right">Load time: {loadTime}ms &ensp; Resolution: {resolution}s &ensp; Result series: {totalTimeSeries}</span>
```

(shorter wording, no title case, and adding a unit to the resolution)

File: web/ui/react-app/src/Panel.tsx [hunk 16]
```diff
@@ -153,14 +154,29 @@ class Panel extends Component<PanelProps, PanelState> {
...
       }
 
+      let totalTimeSeries = 0;
+      if (json.data) {
+        const { resultType, result} = json.data;
```
Comment (juliusv) [inline web/ui/react-app/src/Panel.tsx]: Whitespace consistency nit:

```suggestion
        const { resultType, result } = json.data;
```

File: web/ui/react-app/src/Panel.tsx [hunk 17]
```diff
@@ -153,14 +154,29 @@ class Panel extends Component<PanelProps, PanelState> {
...
+        if (resultType === "scalar") {
+          totalTimeSeries = 1;
+        } else if(result && result.length > 0) {
```
Comment (juliusv) [inline web/ui/react-app/src/Panel.tsx]: Whitespace consistency nit:

```suggestion
        } else if (result && result.length > 0) {
```

**Comment by juliusv** [comment]
Looks great besides nits, thanks :)

File: web/ui/react-app/src/QueryStatsView.tsx (hunk 15 above)
Comment (CSTDev) [reply web/ui/react-app/src/QueryStatsView.tsx]: 🤣 I looked at it and thought about the case, but went with this one. Will change it.

File: web/ui/react-app/src/Panel.tsx (hunk 14 above)
Comment (CSTDev) [reply web/ui/react-app/src/Panel.tsx]: Ah yep good point. Have sorted that.

File: web/ui/react-app/src/QueryStatsView.tsx (hunk 11 above)
Comment (CSTDev) [reply web/ui/react-app/src/QueryStatsView.tsx]: Figured it out, so done it the way you suggested.

**Comment by juliusv** [comment]
:+1: Awesome, looks good now, thanks!
